packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1159 @@
+/// forward and backward buttons.
+///
+/// Otherwise, we strongly recommend implementing the
+/// [RouterDelegate.currentConfiguration] and the [RouteInformationParser.restore]
+/// to provide optimal experience in the web application.
+class Router<T> extends StatefulWidget {
+  /// Creates a router.
+  ///
+  /// The [routeInformationProvider] and [routeInformationParser] can be null if this
+  /// router does not depend on route informations. A common example is a sub router
+  /// that builds its content completely relies on the app state.
+  ///
+  /// If the [routeInformationProvider] is not null, the [routeInformationParser] must
+  /// also not be null.
+  ///
+  /// The [routerDelegate] must not be null.
+  const Router({
+    Key key,
+    this.routeInformationProvider,
+    this.routeInformationParser,
+    this.backButtonDispatcher,
+    @required this.routerDelegate,
+  }) : assert(routeInformationProvider == null || routeInformationParser != null),
+       assert(routerDelegate != null),
+       super(key: key);
+
+  /// The route information provider for the router.
+  ///
+  /// The value at the time of first build will be used as the initial route.
+  /// The [Router] listens to this provider and rebuilds with new names when
+  /// it notifies.
+  ///
+  /// This can be null if this router does not rely on the route information
+  /// to build its content. In such case, the [routeInformationParser] can also be
+  /// null.
+  final RouteInformationProvider routeInformationProvider;
+
+  /// The back button dispatcher for the router.
+  ///
+  /// The two common alternatives are the [RootBackButtonDispatcher] for root
+  /// router, or the [ChildBackButtonDispatcher] for other routers.
+  final BackButtonDispatcher backButtonDispatcher;
+
+  /// The route information parser for the router.
+  ///
+  /// When the [Router] gets a new route information from the [routeInformationProvider],
+  /// the [Router] uses this delegate to parse the route information and produce a
+  /// configuration. The configuration will be used by [routerDelegate] and
+  /// eventually rebuilds the [Router] widget.
+  ///
+  /// Since this delegate is the primary consumer of the [routeInformationProvider],
+  /// it must not be null if [routeInformationProvider] is not null.
+  final RouteInformationParser<T> routeInformationParser;
+
+  /// The router delegate for the router.
+  ///
+  /// This delegate consumes the configuration from [routeInformationParser] and
+  /// builds a navigating widget for the [Router].
+  ///
+  /// It is also the primary respondent for the [backButtonDispatcher]. The
+  /// [Router] relies on the [RouterDelegate.popRoute] to handles the back
+  /// button intends.
+  ///
+  /// If the [RouterDelegate.currentConfiguration] returns a non-null object,
+  /// this [Router] will opt for URL updates.
+  final RouterDelegate<T> routerDelegate;
+
+  /// Retrieve the router of exact type.
```

Comment:
Please see the style guide regarding requirements for documentation. There are writing prompt in the style guide for ideas of things that you should include. The most important idea, though, is that when you're writing the documentation, you have to think of why people might be reading it. They are reading the documentation because they have a question. What is the question? That's what we need to answer.

Reply:
(for what it's worth, the question i had was "why would i use this"; looks like the answer is something to do with the back button dispatcher)